staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go

```diff
@@ -176,6 +176,29 @@ var (
 			Help:           "Counter for consistent reads from cache.",
 			StabilityLevel: compbasemetrics.ALPHA,
 		}, []string{"resource", "success", "fallback"})
+
+	WatchDispatchEventNumWatchers = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "watch_dispatch_event_num_watchers",
+			Help:           "Histogram of total number of watchers to dispatch event to.",
+			Buckets:        []float64{1, 2, 3, 4, 6, 8, 10, 15, 20, 30, 40, 50, 100, 200, 300, 500, 800, 1000, 2000, 3000},
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"resource"},
```

Comment:
Why do we need a histogram here? Wouldn't it be better with a Gauge?

Reply:
This metrics is added to make it easier to inspect the percentile of how many cache watchers are sent per event to detect the possible bottleneck. Say if more than 50% events are sent to 50 or more cache watchers and the write qps to apiserver is more than 1000/s, it will be a lot of event delivery pressure to apiserver.

Reply:
I see, I originally thought the number of watchers to dispatch event to would always be the same depending on the resource, but looking at https://github.com/kubernetes/kubernetes/blob/113c977d9e77001f7f999ddc98a9479fe0162974/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1264-L1287 it sounds more dynamic than I expected, so having a Histogram looks good to me.